SeleniumTunnel.js

```diff
@@ -0,0 +1,229 @@
+			return this._download(config, forceDownload);
+		}.bind(this));
+		
+		return Promise.all(tasks);
+	},
+
+	_download: function (config, forceDownload) {
+		var executable = config.executable;
+		var isDownloaded = fs.existsSync(pathUtil.join(this.directory, executable));
+		var self = {
+			isDownloaded: isDownloaded,
+			url: config.url,
+			proxy: this.proxy,
+			directory: this.directory,
+			executable: executable,
+			dontExtract: !!config.dontExtract
+		};
+		// Leverages the parent download() method by binding a call using the parameters created above
+		// TODO we should consider extracting download into a command/utility function external to Tunnel
+		return _super.download.call(self, forceDownload);
+	},
+	
+	_makeArgs: function () {
+		var directory = this.directory;
+		var seleniumConfig = new descriptors.SeleniumConfig(this.seleniumVersion);
+		var driverConfigs = this._getDriverConfigs();
+		var args = [
+			'-jar',
+			pathUtil.join(this.directory, seleniumConfig.executable),
+			'-port',
+			this.port
+		];
+		
+		driverConfigs.reduce(function (args, config) {
+			var file = pathUtil.join(directory, config.executable);
+			args.push('-D' + config.seleniumProperty + '=' + file);
+			return args;
+		}, args);
+
+		if (this.seleniumArgs) {
+			args = args.concat(this.seleniumArgs);
+		}
+
+		if (this.verbose) {
+			args.push('-debug');
+			console.log('starting with arguments: ', args.join(' '));
+		}
+		
+		return args;
+	},
+
+	_start: function () {
+		function clearHandles() {
+			handles.forEach(function (handle) {
+				handle.remove();
+			});
+		}
+
+		var self = this;
+		var childHandle = this._makeChild();
+		var child = childHandle.process;
+		var dfd = childHandle.deferred;
+		var handles = [
+			util.on(child.stderr, 'data', function (data) {
+				// Selenium recommends that we poll the hub looking for a status response
+				// https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/7957
+				// We're going against the recommendation here for a few reasons
+				// 1. There's no default pid or log to look for errors to provide a specific failure
+				// 2. Polling on a failed server start could leave us with an unpleasant wait
+				// 3. Just polling a selenium server doesn't guarantee it's the server we started
+				// 4. This works pretty well
+				if (data.indexOf('java.net.BindException') > -1) {
+					clearHandles();
+					dfd.reject(new Error(data));
+				}
+				if (data.indexOf('Selenium Server is up and running') > -1) {
+					clearHandles();
+					dfd.resolve();
+				}
+				if (self.verbose) {
+					console.log(data);
+				}
+			}),
+			util.on(child, 'error', function (error) {
+				clearHandles();
+				dfd.reject(error);
+			})
+		];
+
+		return dfd.promise.then(clearHandles, clearHandles)
+			.then(function () {
+				return self._assertStarted()
+					.then(function () {
+						return childHandle;
+					});
+			});
+	},
+
+	_assertStarted: function () {
```

Comment:
From #4 above. They can check to see if selenium is already running on the port.